numpy/core/include/numpy/npy_cpu.h

```diff
@@ -115,4 +116,84 @@
 #define NPY_CPU_HAVE_UNALIGNED_ACCESS 0
 #endif
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum
+{
+    NPY_CPU_FEATURE_NONE = 0,
+    /* X86 */
+    NPY_CPU_FEATURE_MMX               = 1,
+    NPY_CPU_FEATURE_SSE               = 2,
+    NPY_CPU_FEATURE_SSE2              = 3,
+    NPY_CPU_FEATURE_SSE3              = 4,
+    NPY_CPU_FEATURE_SSSE3             = 5,
+    NPY_CPU_FEATURE_SSE41             = 6,
+    NPY_CPU_FEATURE_POPCNT            = 7,
+    NPY_CPU_FEATURE_SSE42             = 8,
+    NPY_CPU_FEATURE_AVX               = 9,
+    NPY_CPU_FEATURE_FP16C             = 10,
+    NPY_CPU_FEATURE_FMA3              = 11,
+    NPY_CPU_FEATURE_FMA               = 11, /* alias to provide backward compatibility */
+    NPY_CPU_FEATURE_XOP               = 12,
+    NPY_CPU_FEATURE_FMA4              = 13,
+    NPY_CPU_FEATURE_AVX2              = 14,
+
+    NPY_CPU_FEATURE_AVX512F           = 30,
+    NPY_CPU_FEATURE_AVX512CD          = 31,
+    NPY_CPU_FEATURE_AVX512ER          = 32,
+    NPY_CPU_FEATURE_AVX512PF          = 33,
+    NPY_CPU_FEATURE_AVX5124FMAPS      = 34,
+    NPY_CPU_FEATURE_AVX5124VNNIW      = 35,
+    NPY_CPU_FEATURE_AVX512VPOPCNTDQ   = 36,
+    NPY_CPU_FEATURE_AVX512BW          = 37,
+    NPY_CPU_FEATURE_AVX512DQ          = 38,
+    NPY_CPU_FEATURE_AVX512VL          = 39,
+    NPY_CPU_FEATURE_AVX512IFMA        = 40,
+    NPY_CPU_FEATURE_AVX512VBMI        = 41,
+    NPY_CPU_FEATURE_AVX512VNNI        = 42,
+    NPY_CPU_FEATURE_AVX512VBMI2       = 43,
+    NPY_CPU_FEATURE_AVX512BITALG      = 44,
+
+    /* X86 CPU Groups */
+    /*(F,CD)*/
+    NPY_CPU_FEATURE_AVX512_COMMON     = 100,
+    /* Knights Landing (F,CD,ER,PF)*/
+    NPY_CPU_FEATURE_XKNL              = 101,
+    /* Knights Mill    (F,CD,ER,PF,4FMAPS,4VNNIW,VPOPCNTDQ) */
+    NPY_CPU_FEATURE_XKNM              = 102,
+    /* Skylake-X       (F,CD,BW,DQ,VL) */
+    NPY_CPU_FEATURE_XSKX              = 103,
+    /* Cannon Lake     (F,CD,BW,DQ,VL,IFMA,VBMI) */
+    NPY_CPU_FEATURE_XCNL              = 104,
+    /* Cascade Lake    (F,CD,BW,DQ,VL,IFMA,VBMI,VNNI) */
+    NPY_CPU_FEATURE_XCEL              = 105,
+    /* Ice Lake        (F,CD,BW,DQ,VL,IFMA,VBMI,VNNI,VBMI2,BITALG,VPOPCNTDQ) */
+    NPY_CPU_FEATURE_XICL              = 106,
+
+    /* IBM/POWER VSX */
+    /* POWER7 */
+    NPY_CPU_FEATURE_VSX               = 150,
+    /* POWER8 */
+    NPY_CPU_FEATURE_VSX2              = 151,
+    /* POWER9 */
+    NPY_CPU_FEATURE_VSX3              = 152,
+
+    /* ARM */
+    NPY_CPU_FEATURE_NEON              = 200,
+    /*(FMA, FP16)*/
+    NPY_CPU_FEATURE_VFPV4             = 201,
+
+    NPY_CPU_FEATURE_MAX               = 256
+};
+
+void npy_cpu_init(void);
+int  npy_cpu_have(int feature_id);
+
+#define NPY_CPU_HAVE(FEATURE_NAME) npy_cpu_have(NPY_CPU_FEATURE_##FEATURE_NAME)
+
+#ifdef __cplusplus
+}
+#endif
```

Comment:
This new code is only used inside `numpy/core/common`, right? If not required, it would be better to move this to a header there

Reply:
On the other hand, it would be nice to allow use from other libraries like scipy

Reply:
> This new code is only used inside numpy/core/common, right? If not required, it would be better to move this to a header there

no, it has actually been used under the hood in every where by gh-13516. e.g. umath genreator and _simd module.

> On the other hand, it would be nice to allow use from other libraries like scipy

Definitely yes.

Reply:
Lesson learned from `numpy.random` refactor: I think it would be prudent to start with this not in the public headers for the first release. The `_simd` module can reach into numpy internals to get the headers, and `umath_generator` generates code, it does not compile it, so I don't see how they require public headers.

Once this is merged and we have a clear picture of how all the parts work together, we can figure out what needs to be publicly exposed.

Reply:
It makes sense to me. thanks!